matview.c

```diff
@@ -1526,6 +1587,161 @@ rewrite_query_for_distinct_and_aggregates(Query *query, ParseState *pstate)
+rewrite_exists_subquery_walker(Query *query, Node *node, int *count)
+{
+	/* This can recurse, so check for excessive recursion */
+	check_stack_depth();
+
+	switch (nodeTag(node))
+	{
+		case T_Query:
+			{
+				FromExpr *fromexpr;
+
+				/* get subquery in WHERE clause */
+				fromexpr = (FromExpr *)query->jointree;
+				query = rewrite_exists_subquery_walker(query, fromexpr->quals, count);
+				/* drop subquery in WHERE clause */
+				if (IsA(fromexpr->quals, SubLink))
+					fromexpr->quals = NULL;
+				break;
+			}
+		case T_BoolExpr:
+			{
+				BoolExprType type;
+
+				type = ((BoolExpr *) node)->boolop;
+				switch (type)
+				{
+					ListCell *lc;
+					case AND_EXPR:
+						foreach(lc, ((BoolExpr *)node)->args)
+						{
+							Node *opnode = (Node *)lfirst(lc);
+							query = rewrite_exists_subquery_walker(query, opnode, count);
+							/* overwrite SubLink node if it is contained in AND_EXPR */
+							if (IsA(opnode, SubLink))
+								lfirst(lc) = makeConst(BOOLOID, -1, InvalidOid, sizeof(bool), BoolGetDatum(true), false, true);
+						}
+						break;
+					case OR_EXPR:
+					case NOT_EXPR:
+						if (checkExprHasSubLink(node))
+							ereport(ERROR,
+									(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+									 errmsg("this query is not allowed on incrementally maintainable materialized view"),
+									 errhint("OR or NOT conditions and EXISTS condition are not used together")));
+						break;
+				}
+				break;
+			}
+		case T_SubLink:
+			{
+				char aliasName[NAMEDATALEN];
+				char columnName[NAMEDATALEN];
+				Query *subselect;
+				ParseState *pstate;
+				RangeTblEntry *rte;
+				RangeTblRef *rtr;
+				Alias *alias;
+				Oid opId;
+				ParseNamespaceItem *nsitem;
+
+				TargetEntry *tle_count;
+				FuncCall *fn;
+				Node *fn_node;
+				Expr *opexpr;
+
+				SubLink *sublink = (SubLink *)node;
+				/* raise ERROR if there is non-EXISTS sublink */
+				if (sublink->subLinkType != EXISTS_SUBLINK)
+					ereport(ERROR,
+							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							 errmsg("this query is not allowed on incrementally maintainable materialized view"),
+							 errhint("subquery in WHERE clause only supports subquery with EXISTS clause")));
+
+				subselect = (Query *)sublink->subselect;
+
+				/* raise ERROR if the sublink has CTE */
+				if (subselect->cteList)
+					ereport(ERROR,
+							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							 errmsg("CTE is not supported on incrementally maintainable materialized view")));
+
+				pstate = make_parsestate(NULL);
+				pstate->p_expr_kind = EXPR_KIND_SELECT_TARGET;
+
+				/*
+				 * convert EXISTS subquery into LATERAL subquery in FROM clause.
+				 */
+
+				snprintf(aliasName, sizeof(aliasName), "__ivm_exists_subquery_%d__", *count);
+				snprintf(columnName, sizeof(columnName), "__ivm_exists_count_%d__", *count);
+
+				/* add COUNT(*) for counting rows that meet exists condition */
+				fn = makeFuncCall(list_make1(makeString("count")), NIL, COERCE_EXPLICIT_CALL, -1);
```

Comment:
This count() must be qualified in order to avoid security issues.

Reply:
Thank you. I fixed to use SystemFuncName().